### PR TITLE
Remove joi to shrink module size

### PR DESCRIPTION
**Pull request by chrisprice**
As noted in #343 & #292 the specific version of `joi` is bloating the package size. However upgrading to the latest version would break ES5 compatibility. 

This PR shrinks the module size from **3.12M** to **0.25M** by replacing joi with bespoke implementation using lodash helpers.

N.B. To keep the implementation as simple as possible, these changes mean that the validation error messages are marginally less helpful e.g. -

`"expiresIn" must be a number` versus `"expiresIn" is not the correct type`

This should also solve #334.

### CODE REVIEW

File: sign.js [hunk 1]
```diff
@@ -1,29 +1,52 @@
...
+  nbf: isNumber
+};
 
+function validate(schema, unknown, object) {
+  if (!isPlainObject(object)) {
+    throw new Error('Expected object');
+  }
+  Object.keys(object)
+    .forEach(function(key) {
+      if (schema[key] == null) {
```
Comment (ziluvatar) [inline sign.js]: `!schema[key]` ?

**Comment by chrisprice** [comment]
Changes sound good, as soon as I get a chance I'll update the PR.

File: test/schema.tests.js [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    });
+
+    it('should validate algorithm', function () {
+      expect(function () {
+        sign({ algorithm: 'foo' });
+      }).to.throw(/"algorithm" must be a valid string enum value/);
+      sign({algorithm: 'RS256'});
+      sign({algorithm: 'RS384'});
+      sign({algorithm: 'RS512'});
+      // not sure how to fix error - 'Could not find expected "seq"'
```
Comment (chrisprice) [inline test/schema.tests.js]: Wasn't sure how to get these tests to pass, I assume I'm missing some required configuration for the `ES*` algorithms.

**Comment by chrisprice** [comment]
So it took a little longer than I expect to get back around to doing this but I think I've made all of the suggested changes.

File: test/schema.tests.js (hunk 2 above)
Comment (ziluvatar) [inline test/schema.tests.js]: You need to use different certificates for RS and ES. Probably it's that.
Take a look to: https://github.com/auth0/node-jsonwebtoken/blob/master/test/jwt.asymmetric_signing.tests.js#L13

File: test/iat.tests.js [hunk 3]
```diff
@@ -12,11 +12,4 @@ describe('iat', function () {
     expect(result.exp).to.be.closeTo(iat + expiresIn, 0.2);
   });
 
-
-  it('should throw if iat is not a number', function () {
```
Comment (ziluvatar) [inline test/iat.tests.js]: Why do you remove it? is it because with this PR it is tested in the schema test?

Comment (chrisprice) [inline test/iat.tests.js]: Yep, that's correct.

**Comment by chrisprice** [comment]
Thanks for the pointer on ECDSA signing, I've updated the tests to use the other key where appropriate.

**Pull request merged**